Add tiered policies to Calico OSS (#1699)
tigera/docs: calico-enterprise/network-policy/policy-tiers/tiered-policy.mdx

- Network and global network policies are placed when you upgrade from Project Calico to {{prodname}}
- Recommended policies are placed when you use the **Recommend a policy** feature

## AdminNetworkPolicy tier:

> The AdminNetworkPolicy tier where all [Kubernetes admin network policies](https://network-policy-api.sigs.k8s.io/reference/examples/) reside. It is automatically created during installation and has the order of 1,000. This value is fixed, and cannot be changed.

Reviewer: ```suggestion
The AdminNetworkPolicy tier is where all [Kubernetes admin network policies](https://network-policy-api.sigs.k8s.io/reference/examples/) reside. It is automatically created during installation and has the order of 1,000. This value is fixed, and cannot be changed.
```

Author: Fixed.

- Network and global network policies are placed when you upgrade from Project Calico to {{prodname}}
- Recommended policies are placed when you use the **Recommend a policy** feature

## AdminNetworkPolicy tier:

> The AdminNetworkPolicy tier is created during installation and has the order of 1,000.
> 
> The AdminNetworkPolicy tier is where:
> 
> - You manage all Kubernetes admin network policies
> 

Reviewer: ```suggestion
The AdminNetworkPolicy tier where is all Kubernetes admin network policies reside. It is automatically created during installation and has the order of 1,000.

```

Reviewer: I think there should be a link to kANP documentation!

## Tier order

> Tiers are ordered based on their orders (from left to right), starting from the lowest order (the highest priority) tiers.

Reviewer: ```suggestion
Tiers are sorted by their orders (from left to right), starting from the lowest order (the highest priority) tiers.
```

Author: Fixed.

- **security tier** - is higher priority than platform tier
- **platform tier** - is higher priority than default tier

> - **default tier** - is the lowest priority than the rest

Reviewer: ```suggestion
- **default tier** - is the lowest priority.
```

Previously, is could not be reordered, now it has an order. Can it be reordered? If yes, it would be worth to say that "previously it could not be reordered. It can be now, but we advise against that" or something similar?

Author: No, the default tier **must** have the order of 1M. It cannot be reordered. Previously, it had an order of `nil`, and now it is `1M` and it cannot change. Also any  tier with `nil` order will be treated as `1M`.
We will add a release note to warn customers about the change.

### Implicit default deny

As shown in the following diagram, at the end of each tier is an implicit default deny. This is a safeguard that helps mitigate against unsecured policy.

> To override the implicit default deny, and to allow traffic evaluation to continue, you can change tier's default action to **Pass**.

Reviewer: I think some of the original text should remain in some form:

```
 Because of this safeguard, you must explicitly apply the **Pass** action rule when you want traffic evaluation to continue. In the following example, the Pass action in a policy ensures that traffic evaluation continues, and overrides the implicit default deny.
 ```
 
 This is still somewhat true as this is one way how to get from tier to another. Now we are adding another method, which may not be as safe. Btw. the ANP tier has pass as a default action? I would think so.

Author: Yes, ANP tier has `Pass` default action.

Kept the original text, and added the option with the default action.

To override the implicit default deny, and to allow traffic evaluation to continue, you can change tier's default action to **Pass**.

![implicit-deny](/img/calico-enterprise/implicit-deny.svg)

> Let’s look at a Dev/Ops global network policy in a high precedence tier (Platform). The policy denies ingress and egress traffic to workloads that match selector, `env == "stage"`. To ensure that policies continue to evaluate traffic after this policy, the policy adds a Pass action for both ingress and egress.

Reviewer: It feels like this should remain

Author: Fixed.

apiVersion: projectcalico.org/v3
kind: Tier
metadata:
  name: security
spec:
  order: 300

>   defaultAction: Pass

Reviewer: And this should be in addition to the pass rule. We did not remove the pass rule, right?

Author: No, with the `defaultAction` added, there is no need to add rules with Pass action just to prevent hitting the implicit deny. I think the previous method looked like a hack, and this version is cleaner.

Reviewer: I understand that we have a better approach, but I mean we have two way now. At the least, We should add that the previous method still works, but is not recommended. There are people who will upgrade and will not change their rules and may get confused if something does not work, that they are doing it wrong. WDYT?

Author: It makes sense to document both options. Adding a pass rule is more specific than having default pass action.
